### PR TITLE
fix: add filter in hnsw searcher

### PR DESCRIPTION
**Pull request by numb3r3**

### CODE REVIEW

File: jinahub/indexers/searcher/HnswlibSearcher/tests/unit/test_searcher.py [hunk 1]
```diff
@@ -80,6 +80,12 @@ def test_index():
     assert index._index.element_count == NUM_DOCS
     assert set(index._ids_to_inds.keys()) == set(da1.get_attributes('id'))
 
+    da3 = DocumentArray([Document() for emb in embeddings])
```
Comment (tadejsv) [inline jinahub/indexers/searcher/HnswlibSearcher/tests/unit/test_searcher.py:83]: Can you move this out to a separate test function? I prefer having each unit test as atomic as possible

File: jinahub/indexers/searcher/HnswlibSearcher/hnswlib_searcher.py [hunk 2]
```diff
@@ -160,7 +160,9 @@ def index(
         if docs is None:
             return
 
-        docs_to_update = docs.traverse_flat(traversal_paths)
+        docs_to_update = docs.traverse_flat(
```
Comment (cristianmtr) [inline jinahub/indexers/searcher/HnswlibSearcher/hnswlib_searcher.py:163]: As agreed, make this a parameter to control whether or not this filter is applied

File: jinahub/indexers/searcher/HnswlibSearcher/tests/unit/test_searcher.py (hunk 1 above)
Comment (numb3r3) [reply jinahub/indexers/searcher/HnswlibSearcher/tests/unit/test_searcher.py]: addressed in latest commits

File: jinahub/indexers/searcher/HnswlibSearcher/hnswlib_searcher.py [hunk 3]
```diff
@@ -29,6 +29,7 @@ def __init__(
         dump_path: Optional[str] = None,
         traversal_paths: Iterable[str] = ('r',),
         is_distance: bool = True,
+        ignore_invalid_docs: bool = True,
```
Comment (nan-wang) [inline jinahub/indexers/searcher/HnswlibSearcher/hnswlib_searcher.py:32]:
```suggestion
        check_embedding: bool = True,
```

File: jinahub/indexers/searcher/HnswlibSearcher/hnswlib_searcher.py [hunk 4]
```diff
@@ -46,6 +47,7 @@ def __init__(
         :param traversal_paths: The default traverseal path on docs (used for indexing,
             search and update), e.g. ['r'], ['c']
         :param is_distance: Boolean flag that describes if distance metric need to be reinterpreted as similarities.
+        :param ignore_invalid_docs: Boolean flat that determine whether filter out invalid docs (i.e., without embedding)
```
Comment (nan-wang) [inline jinahub/indexers/searcher/HnswlibSearcher/hnswlib_searcher.py:50]:
```suggestion
        :param check_embedding: If `check_embedding` is False (default: True), then the validation of embedding   will be skipped. Documents without embeddings will result in an `TypeError` or `ValueError`.
```

**Review by nan-wang — CHANGES_REQUESTED** [review]

File: jinahub/indexers/searcher/HnswlibSearcher/hnswlib_searcher.py (hunk 3 above)
Comment (numb3r3) [reply jinahub/indexers/searcher/HnswlibSearcher/hnswlib_searcher.py:32]: the parameter name`check_embedding=True` cannot indicate the action of filtering.  Could the users confuse this parameter?

Comment (cristianmtr) [reply jinahub/indexers/searcher/HnswlibSearcher/hnswlib_searcher.py:32]: I'd propose `filter_invalid_docs`

File: jinahub/indexers/searcher/HnswlibSearcher/hnswlib_searcher.py [hunk 5]
```diff
@@ -164,7 +167,19 @@ def index(
         if len(docs_to_update) == 0:
             return
 
-        embeddings = docs_to_update.embeddings
+        try:
+            embeddings = docs_to_update.embeddings
+        except (TypeError, ValueError) as ex:
+            if self.ignore_invalid_docs:
+                docs_to_update = docs_to_update.traverse_flat(
+                    ['r'], filter_fn=lambda d: d.embedding is not None
```
Comment (cristianmtr) [inline jinahub/indexers/searcher/HnswlibSearcher/hnswlib_searcher.py:175]: and `d.embedding.shape == self.blablabla.shape`, no?

Comment (numb3r3) [reply jinahub/indexers/searcher/HnswlibSearcher/hnswlib_searcher.py:175]: I would say the mis-matched shape can directly raise an exception.

Comment (cristianmtr) [reply jinahub/indexers/searcher/HnswlibSearcher/hnswlib_searcher.py:175]: Yesterday we agreed that Indexer should not implement any filtering. It is the job of the upstream to make sure documents are correct. So this PR should be closed, right?

Comment (numb3r3) [reply jinahub/indexers/searcher/HnswlibSearcher/hnswlib_searcher.py:175]: yes, we can close this PR.

**Comment by numb3r3** [comment]
The filter should be done in upstream.

**Pull request closed without merging**